### DISCUSSION TITLE
Sensor with high RPM creates an intractable action space

### DISCUSSION THREAD

**Discussion by iwishiwasaneagle** [dstl/Stone-Soup#1089] [Q&A] [answered]
The current implementations of the various action generators creates an impossibly large action space for the sensor managers to handle. I am wanting to implement a high-RPM model with a continuous controller. How would this be possible?

**Comment by akenyon** [comment, marked as answer]
Hi @iwishiwasaneagle thanks for raising this.

This may be something we'd want to add as a feature in the future, but as a quick workaround you could create a slightly modified subclass of the `DwellActionsGenerator`. You could overwrite the `angle_delta` property to restrict the actions generated to those within a single rotation of the sensor.

```python
import numpy as np

from stonesoup.sensor.action.dwell_action import DwellActionsGenerator
from stonesoup.types.angle import Angle

class CustomDwellActionsGenerator(DwellActionsGenerator):
    @property
    def angle_delta(self):
        angle = Angle(self.duration.total_seconds() * self.rps * 2 * np.pi)
        return angle if angle < np.pi else np.pi
```

This would cap the number of actions generated at $2\pi/resolution$, rather than $rotation speed*2\pi/resolution$.

Then to get your sensor of choice to use the custom action generator, you could subclass it and change the `generator_cls` of the `dwell_centre` variable. E.g. for `RadarRotatingBearingRange`:

```python
from stonesoup.sensor.radar import RadarRotatingBearingRange
from stonesoup.sensormanager.action import ActionableProperty
from stonesoup.types.state import StateVector

class CustomRadarRotatingBearingRange(RadarRotatingBearingRange):
    dwell_centre: StateVector = ActionableProperty(
        doc="A `state_vector` property that describes the rotation angle of the centre of the "
            "sensor's current FOV (i.e. the dwell centre) relative to the positive x-axis of the "
            "sensor frame/orientation. The angle is positive if the rotation is in the "
            "counter-clockwise direction when viewed by an observer looking down the z-axis of "
            "the sensor frame, towards the origin. Angle units are in radians",
        generator_cls=CustomDwellActionsGenerator,  # this is the only line changed from the original
        generator_kwargs_mapping={'rpm': 'rpm', 'resolution': 'resolution'})
```

Hope this helps.

Alex

**Reply by iwishiwasaneagle** [reply]
This worked great, thanks! Might be worth putting a warning in place if RPM is high enough to suffer from floating point errors?
